New module: xenserver_guest - manages XenServer VMs

bvitnik opened the pull request:
##### SUMMARY
Ansible currently lacks any support for managing XenServer related objects and resources. This module implements management of XenServer VMs (guests) with module interface comparable to vmware_guest module. In simple use cases, these modules are interchangeable. The pull request also brings new module util that xenserver_guest is dependent on. Module util is meant to be a basis for future XenServer related modules. At least one more module using this module util is planed in near future - xenserver_guest_facts - a simple fact gathering only module.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
xenserver_guest

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (xenserver 12d0febe14) last updated 2018/06/02 02:21:44 (GMT +200)
  config file = /home/bvitnik/.ansible.cfg
  configured module search path = [u'/home/bvitnik/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /srv/bvitnik/projects/ansible/ansible/lib/ansible
  executable location = /srv/bvitnik/projects/ansible/ansible/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```


##### ADDITIONAL INFORMATION
- ~~The module currently support most of the features of vmware_guest module. Notable lacking feature is guest OS customization because of the lack of proper support on a platform level. There is limited official support for network configuration as of XenServer 7.0 but only for Windows guests using official XenServer agent:~~

  ~~https://xenserver.org/blog/entry/set-windows-guest-vm-static-ip-address-in-xenserver.html~~

  ~~This will be supported by the module in the future. I'm currently also exploring alternatives. One of the alternatives would be to support OpenStack nova-agent for XenServer guests (developed by Rackspace) but it seems to be unmaintained in recent years. Another alternative would be to expose network configuration through xenstore and just leave it up to user to implement custom agent/script inside a guest that does all the network and OS customization.~~

  Guest OS customization of network parameters has now been implemented.

Commit:
XenServer related modules - initial commit

 - New module_util: xenserver. Contains common module arguments, functions
   and classes useful for future XenServer related modules.
 - New module_docs_fragment: xenserver. Describes common module arguments.
 - New module: xenserver_guest. Supports VM deployment, reconfiguration,
   removal, detection of changes, state management, fact gathering and
   Ansible check mode. Module is fully documented.
 - Updated: developing_module_utilities.rst.
 - Module params, workflow and some functions are based on or taken from
   vmware_guest module.

Commit:
Implemented support for configuring custom VM params in xenserver_guest module

Commit:
Compatibility fixes and documentation update

 - xenserver module_util: implemented support for XenAPI.py version older
   than 7.2.
 - xenserver module_util: PEP8 fixes.
 - xenserver module_util: Added missing imports.
 - xenserver module_util: Copyright notice fixes.
 - xenserver_guest module: updated module documentation with notes regarding
   module requirements and compatibility.
 - xenserver_guest module: bumped version_added to 2.7.
 - xenserver_guest module: minor fixes.

Akasurde commented on lib/ansible/module_utils/xenserver.py:
```diff
@@ -0,0 +1,300 @@
+# -*- coding: utf-8 -*-
+
+# (c) 2018, Bojan Vitnik <bvitnik@mainstream.rs>
```
`Copyright: (c) 2018, Bojan Vitnik <bvitnik@mainstream.rs>`

Akasurde commented on lib/ansible/module_utils/xenserver.py:
```diff
@@ -0,0 +1,300 @@
+
+# (c) 2018, Bojan Vitnik <bvitnik@mainstream.rs>
+#
+# This file is part of Ansible
```
Replace lines 5 to 18 with

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

Akasurde commented on lib/ansible/modules/cloud/xenserver/xenserver_guest.py:
```diff
@@ -0,0 +1,1510 @@
+notes:
+- Minimal supported version of XenServer is 5.6
+- Module was tested on XenServer 6.5, 7.1 and 7.2
+- To use https:// scheme for C(hostname) and C(validate_certs=no), XenAPI.py from XenServer 7.2 SDK or newer and pyhon 2.7.9 or newer are needed
```
s/pyhon/Python/

Akasurde commented on lib/ansible/module_utils/xenserver.py:
```diff
@@ -0,0 +1,300 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
```
Add - 

```
from __future__ import absolute_import, division, print_function
__metaclass__ = type
```

Akasurde commented on lib/ansible/modules/cloud/xenserver/xenserver_guest.py:
```diff
@@ -0,0 +1,1510 @@
+            if vm_power_state == "running" and self.module.check_mode:
+                return True
+
+            # running state is required for guest reboot
```
Could you please take a look at sub argument spec mentioned in https://github.com/ansible/ansible/pull/27119#issue-131585030 ? it would easy for user input validation and module simplification.

Akasurde replied:
Same for cdrom, custom_params and networks

bvitnik replied:
This looks great. I'll check into it. I wondered why such functionality was missing.

bvitnik replied:
Is this a requirement for merging the module to Ansible code base or just a suggestion for future development? This would require some rethinking of the code workflow and a lot of testing.

Akasurde replied:
Sub argument is not requirement but will make code concise and remove additional code used for user input.

bvitnik replied:
Definitely. This will be high on my TODO list but how soon I will be able to implement it and test remains to be seen.

Akasurde commented on lib/ansible/modules/cloud/xenserver/xenserver_guest.py:
```diff
@@ -0,0 +1,1510 @@
+            if vm_power_state == "running" and self.module.check_mode:
+                return True
+
+            # running state is required for guest reboot
```
I would recommend to have a separate module for managing power state of VMs. We have seen issues in vmware_guest module related to this and created new module called vmware_guest_powerstate ?

bvitnik replied:
Hmmm. If that is a general course of action with vmware_guest module, I can go with that too. So is power state management for vmware_guest module planned for removal? Also, if I split power state management to separate module, should I make that module part of this pull request?

Akasurde replied:
There is no plan to remove power state management from vmware_guest for backward compatibility reason. But, for this module I would say it will be good to manage powerstate separately. Submit a separate power state module as per - https://docs.ansible.com/ansible/2.5/dev_guide/developing_modules_in_groups.html#subsequent-prs

bvitnik replied:
Hmmm, with current design a user can specify power state "poweredon" for non existing VM. In that case, the module will deploy a new VM based on module parameters and run it automatically after deploy. I wouldn't like for users to be forced to use two separate modules and two tasks in a playbook to accomplish the same, especially if they can do the same with vmware_guest module and will be able to do it for foreseeable future. I think that at least "poweredon" and maybe "poweredoff" states should be supported by this module - it removes the clutter from playbook. Other states are much less relevant.

Akasurde replied:
ok.

bvitnik replied:
Great. Then my plan is to split power state management functionality from xenserver_guest module and put it in xenserver module util so that it can be shared by multiple modules.

Akasurde commented:
Does Xenserver provides any facility to emulate its APIs ? For vmware_* modules we have https://github.com/ansible/vcenter-test-container 

I would love to see such facility for this module as well. This will reduce number of issues and will make sure new changes will not break existing functionality.

bvitnik commented:
I would love to have XenAPI simulator too but I'm not aware of existence of such a thing. I'll try asking around.

Akasurde commented:
@bvitnik Thanks for the PR.

bvitnik commented:
@Akasurde Thanks for all the comments.

Commit:
VM power state management refactoring, subargument specs in xenserver_guest module, other fixes

 - VM power state management code moved from xenserver_guest module to xenserver
   module_util (set_vm_power_state function).
 - Code for waiting for VM IP address moved from xenserver_guest module to
   xenserver module_util (wait_for_vm_ip_address function).
 - xenserver module_util: implemented get_object_ref function to clean up
   a lot of repeated code in xenserver_guest module.
 - xenserver module_util: added additional aliases 'host' and 'pool' for
   'hostname' common module argument. They are more in line with what
   XenServer users are familiar with.
 - xenserver module_util: minor fixes.
 - xenserver_guest module: removed VM state management other than 'present',
   'absent' and 'poweredon'. Other states are to be managed by separate module.
 - xenserver_guest module: added subargument specs and cleaned up custom code
   for subargument validation.
 - xenserver_guest module: reorganized code for disk and network reconfiguation
   to minimize code duplication.
 - xenserver_guest module: renamed 'cdrom.iso' module argument to
   'cdrom.iso_name', avoids cryptic error message when parameter is
   missing.
 - xenserver_guest module: documentation update.
 - xenserver_guest module: changes in error messages.
 - xenserver_guest module: minor fixes.

bvitnik commented:
@Akasurde Sorry for the delay. I've moved VM power state management to module util so that I can base new xenserver_guest_powerstate module on it. I've kept support for "poweredon" state in the module. I couldn't find real use case for keeping "poweredoff" state so I removed it along side other power states. Should there be any demand from users to support all power states directly in xenserver_guest module, it can be trivially brought back.

I've also defined subargument specs as you suggested. It works OK and does some validation better than my original code. On the other hand, some error messages are rather... hmmm... cryptic :), like:

> "msg": "type is iso but all of the following are missing: iso found in cdrom"

I will implement xenserver_guest_facts and xenserver_guest_powerstate modules soon as they are now trivial to implement. These will be separate pull requests.

I also plan to implement some rudimentary support for IP/gateway OS customization soon. My idea is to try to detect if official agent (XenServer Guest Tools) supports IP/gateway OS customization and then use it. If not, the module can fallback to writing IP/gateway info to xenstore which can be accessed from inside of VM. In that case, users should prepare custom scripts that run at boot time to do all the network configuration. Does that sound reasonable?

Regarding XenAPI simulation, I was assured by developers of Xen Orchestra that such a thing does not exist publicly and that they are pretty confident that Citrix does not have it in house either. So any module testing unfortunately must be done on real hardware or maybe virtualized (nested) environment.

Akasurde commented on lib/ansible/modules/cloud/xenserver/xenserver_guest.py:
```diff
@@ -0,0 +1,1241 @@
+            type='dict',
+            options=dict(
+                type=dict(type='str', required=True, choices=['none', 'host', 'iso']),
+                iso=dict(type='str'),
```
That cryptic is error is coming from here. You can rename this parameter as `iso_image` or something similar which makes easy to understand for end user and will resolve the error too.

bvitnik replied:
Yes, I know where it comes from. My thinking was I don't want to break compatibility with vmware_guest module by renaming the argument but then I realized I already did that - its *iso_path* there and *iso* here (path does not make sense in XenServe). So I will be renaming it to *iso_name* since most objects have a *name*.

Akasurde commented:
> Sorry for the delay.

No problem 


>  I've moved VM power state management to module util so that I can base new xenserver_guest_powerstate module on it. I've kept support for "poweredon" state in the module. I couldn't find real use case for keeping "poweredoff" state so I removed it along side other power states. Should there be any demand from users to support all power states directly in xenserver_guest module, it can be trivially brought back.

Nice.

> I've also defined subargument specs as you suggested. It works OK and does some validation better than my original code. On the other hand, some error messages are rather... hmmm... cryptic :), like:

Commented in code

> I will implement xenserver_guest_facts and xenserver_guest_powerstate modules soon as they are now trivial to implement. These will be separate pull requests.

OK

> I also plan to implement some rudimentary support for IP/gateway OS customization soon. My idea is to try to detect if official agent (XenServer Guest Tools) supports IP/gateway OS customization and then use it. If not, the module can fallback to writing IP/gateway info to xenstore which can be accessed from inside of VM. In that case, users should prepare custom scripts that run at boot time to do all the network configuration. Does that sound reasonable?

Sounds OK to me.

> Regarding XenAPI simulation, I was assured by developers of Xen Orchestra that such a thing does not exist publicly and that they are pretty confident that Citrix does not have it in house either. So any module testing unfortunately must be done on real hardware or maybe virtualized (nested) environment.

If there is nothing like simulator then we have to rely on community with hardware for both testing and reporting stability of module. Anyways that will not become blocker for this module for now.

bvitnik commented:
@Akasurde Thanks for the comments. Do I need to rebase my branch? Also, should I do some commit history cleaning? It's kind of messy in some places. Anyway, is there anything more to do before merging (except renaming *iso* to *iso_name*)? I've also compiled the doc pages and they look nice.

For IP/gateway OS customization support, I suppose it's probably better for it to be separate PR after this PR is merged?

Commit:
Implemented guest OS network parameter configuration and other

 - xenserver module_util: moved code for validating MAC addresses from
   xenserver_guest module and implemented a range of functions for validating
   IP addresses and related entities and converting prefixes to netmasks and
   vice versa.
 - xenserver module_util: updated fact gathering code to support guest OS
   network parameters.
 - xenserver module_util: added docstrings.
 - xenserver module_util: minor changes.
 - xenserver_guest module: implemented support for guest OS network parameter
   configuration.
 - xenserver_guest module: changed CD-ROM handling code.
 - xenserver_guest module: changed so that user friendly version of changes
   list is always returned in module result.
 - xenserver_guest module: error message changes.
 - xenserver_guest module: added docstrings.
 - xenserver_guest module: documentation update.
 - xenserver_guest module: minor changes and fixes.

bvitnik commented:
@Akasurde Hi. Last 3 days have been productive. I've implemented all originally planned features including guest OS network parameter configuration. Code has also been cleaned in a lot of places and thoroughly tested but I'll have to rely on users for further testing. I've also cleaned commit history. This now looks as good as I wanted it to be. Can you take a quick look at the module? If there is nothing more to be changed, as far as I'm concerned, this is ready for merging.

Akasurde commented:
cc @dericcrago @pdellaert @jctanner This module is similar to `vmware_guest` , could you please take a look so that we can merge this ?
Thanks in advance.

maxamillion commented:
FWIW, I think the code looks good. I don't have a Xenserver to test it against though, so take that with a grain of salt. I would like to point out the [recent request for new noteworthy changes to come with a changelog snippet](https://groups.google.com/forum/#!topic/ansible-devel/uYuIADYa5vs) (which I think this qualifies).

bvitnik commented:
@maxamillion Are you sure it qualifies for inclusion of changelog snippet? On a page provided it is said that changelog snippets are needed "except for new modules and plugins (Our tooling 
will automatically add any new module or plugin)". This PR brings new module and module util.

maxamillion commented:
@bvitnik fair point, you are correct. It probably doesn't one, disregard. :)

Commit:
Various fixes and code cleanup

 - xenserver module_util: implemented get_xenserver_version function.
 - xenserver module_util: moved customization agent detection code to
   gather_vm_params function. customization_agent variable is now part of
   vm_params. An exception in customization agent detection code that prevented
   deployment of new VMs is also fixed.
 - xenserver module_util: added support for alternative VM state names with
   dash and underscore in set_power_state function.
 - xenserver_guest module: removed customization agent detection code
   as it is now implemented in xenserver module_util.
 - xenserver_guest module: fixed a bug in xenserver_data update code that
   occured when "networks.mac" was not specified in module params and
   other fixes.
 - xenserver_guest module: some code cleanup.
 - xenserver_guest module: bumped version_added to 2.8.

bvitnik commented:
@Akasurde @dericcrago @pdellaert @jctanner @maxamillion Hi everyone. I've been rather busy last few months so I couldn't spare much time to work on this but I haven't forsaken it. Latest commit fixes some bugs I found in the mean time and code is now based on Ansible 2.8 devel. Since module needs some automated testing and there is no XenAPI simulator in the wild, the idea was for me to try to implement a fake XenAPI provider. I decided to implement a full XML-RPC server to simulate XenAPI provider so that any XenAPI client can connect to it:

https://github.com/bvitnik/XenAPISim/tree/XenAPISim

Implementing an XML-RPC server and defining most classes and methods was an easy part. On the other hand, implementing some of the methods proved to be challenging as deep knowledge of XenServer internals is needed to implement them properly. XenServer and XenAPI source code written in OCaml was of little help (still trying to wrap my head around some OCaml concepts) so the work on it has been on hold for some time. I intend to get back on it at some point. Unfortunately I still haven't received any useful testing feedback, sorry :(, so all the testing is mostly done by me and my colleges at work. That's it for a quick status update.

Akasurde commented:
rebuild_merge

bvitnik commented:
> rebuild_merge

@Akasurde Oh this came as a surprise. Thanks for merging the PR. Is it OK now for me to make pull requests for other two accompanying modules - xenserver_guest_powerstate and xenserver_guest_facts? I have them already implemented.
